cmd/groundstation/main.go

```diff
@@ -1,11 +1,41 @@
 package main
 
 import (
-	"fmt"
-	"github.com/gpuctl/gpuctl/internal/incrementer"
+	"log/slog"
+	"net/http"
+	"time"
+
+	"github.com/gpuctl/gpuctl/internal/femto"
+	"github.com/gpuctl/gpuctl/internal/uplink"
 )
 
 func main() {
-	fmt.Println("Hello from groundstation")
-	fmt.Println(incrementer.Inc(4))
+	slog.Info("Starting groundstation")
+
+	mux := new(femto.Femto)
+	gs := groundstation{
+		lastSeen: make(map[string]time.Time),
+	}
+	femto.OnPost(mux, uplink.HeartbeatUrl, gs.heartbeat)
+
+	// TODO: Make this configurable
+	err := mux.ListenAndServe(":8080")
+
+	slog.Info("Shut down groundstation", "err", err)
+}
+
+type groundstation struct {
+	lastSeen map[string]time.Time
+}
+
+func (gs *groundstation) heartbeat(data uplink.HeartbeatReq, req *http.Request, log *slog.Logger) error {
+	// TODO: Pull out just the IP here.
+	from := req.RemoteAddr
+	prev := gs.lastSeen[from]
```

Comment:
This is racy. But no point fixing until it can be reproduced in a test. Easy enough to reproduce.

```go
toSpawn := 10000
	wg.Add(toSpawn)
	for i := 0; i < toSpawn; i++ {
		go func() {
			err := s.sendHeartBeat()
			if err != nil {
				log.Error("failed to send heartbeat", "err", err)
			}
			wg.Done()
		}()
	}
	wg.Wait()
```

but needs infra to spin up a test server

Reply:
Hey Alona, I agree that this is currently racy, but would I be right in thinking it would be solved once we're putting last seen into the database rather than storing it in a data structure shared by all the request handlers?

Reply:
Yeah, the DB layer should fix this specific instance.

However I still want to land tests for this, as it'll force me to write enough of a harness to write a failing test (with the race detector).

Reply:
Do we want to fix this by making the `lastSeen` map thread-safe, or by building in a thread-safe event queue (or using Go channels)?

Reply:
Current approach: Slap a big mutex over it. Doing event queues is much more fragile.

(ask me about the horrors of the conversation event pump sometime :))